### PR TITLE
govnc builtin proxy

### PR DESCRIPTION
**Pull request by ksimuk**

### CODE REVIEW

File: xserver/playbook.yaml [hunk 1]
```diff
@@ -3,74 +3,76 @@
...
-      add_preview_to_codio:
-    - name: Restore .codio permissions
-      file:
-        path: /home/codio/workspace/.codio
-        owner: codio
-        group: codio
+    - name: Add Virtual Desktop to Preview menu
+      become_user: codio
+      shell: 
+        cmd: 'cat .codio | sed -E -e "s:(\s+(//|#)|^\s*(//|#)).*$::; /^$/d" | jq  ". * {\"preview\": {\"Virtual Desktop\": \"https://{%raw%}{{domain3050}}{%endraw%}/\"}}" > .codio_tmp ; mv .codio_tmp .codio ; true'
```
Comment (destitutus) [inline xserver/playbook.yaml:74]: 1. will it work if `.codio` does not exist?
2. will it handle comments in proper way(.codio) support comments

Comment (ksimuk) [reply xserver/playbook.yaml:74]: 1. if `.codio` file exist it won't add anything
2. yes it handles comments

Comment (destitutus) [reply xserver/playbook.yaml:74]: > if .codio file exist it won't add anything

but it looks like regression, previous code does add new `.codio` file

> yes it handles comments

how will it be done? will they stay where they are?

Comment (ksimuk) [reply xserver/playbook.yaml]: 1. fixed

Comment (ksimuk) [reply xserver/playbook.yaml]: > how will it be done? will they stay where they are?

no, as they were not before

File: xserver/playbook.yaml [hunk 2]
```diff
@@ -3,74 +3,76 @@
...
-      add_preview_to_codio:
-    - name: Restore .codio permissions
-      file:
-        path: /home/codio/workspace/.codio
-        owner: codio
-        group: codio
+    - name: Add Virtual Desktop to Preview menu
+      become_user: codio
+      shell: 
+        cmd: '(cat .codio || echo "{}") | sed -E -e "s:(\s+(//|#)|^\s*(//|#)).*$::; /^$/d" | jq  ". * {\"preview\": {\"Virtual Desktop\": \"https://{%raw%}{{domain3050}}{%endraw%}/\"}}" > .codio_tmp ; mv .codio_tmp .codio ; true'
```
Comment (destitutus) [inline xserver/playbook.yaml:74]: The file will be empty if the syntax is incorrect.
Can we add a Virtual Desktop preview there as before?

File: xserver/playbook.yaml [hunk 3]
```diff
@@ -3,74 +3,76 @@
...
-      add_preview_to_codio:
-    - name: Restore .codio permissions
-      file:
-        path: /home/codio/workspace/.codio
-        owner: codio
-        group: codio
+    - name: Add Virtual Desktop to Preview menu
+      become_user: codio
+      shell: 
+        cmd: '(cat .codio || echo "{}") | sed -E -e "s:(\s+(//|#)|^\s*(//|#)).*$::; /^$/d" | jq  ". * {\"preview\": {\"Virtual Desktop\": \"https://{%raw%}{{domain3050}}{%endraw%}/\"}}" .codio_tmp && mv .codio_tmp .codio ; true'
```
Comment (destitutus) [inline xserver/playbook.yaml:74]: did you miss `>` ? in `endraw%}/\"}}" .codio_tmp`

**Review by destitutus — APPROVED** [review]

**Pull request merged**